File: domains/org/src/DTOs/Team.cs

namespace Org.DTOs;

public class TeamCollectionDto
{
  public int Id { get; set; }
  public string Name { get; set; } = null!;
  public string? Description { get; set; }
  public string Type { get; set; } = null!;
  public int? ParentId { get; set; }
  public List<TeamDto> Children { get; set; } = new();
  public DateTimeOffset FormedAtDate { get; set; }
  public DateTimeOffset? DisbandedAtDate { get; set; }
}

public class TeamDto : TeamCollectionDto
{
  public TeamDto? Parent { get; set; }
}

public class CreateTeamDto : UpdateTeamDto
{
  public required string Type { get; set; }
  public DateTimeOffset FormedAtDate { get; set; } = DateTimeOffset.UtcNow;
}

public class UpdateTeamDto
{
  public required string Name { get; set; }
  public string? Description { get; set; }
  public int? ParentId { get; set; }
  public DateTimeOffset? DisbandedAtDate { get; set; }
}

File: webapp/src/Client/Team.ts
import { useCRUD, UseCRUDOptionalProps } from "./CRUD/CRUD";

export interface Team {
  id: number;
  name: string;
  description?: string;
  type: string;
  formed_at_date: string;
  disbanded_at_date?: string;
  parent_id?: number;
  parent?: Team;
  children: Team[];
}

export interface MutateTeam {
  name: string;
  description?: string;
  type: string;
  parent_id?: number;
  formed_at_date: string;
  disbanded_at_date?: string;
}

export interface UseTeamProps extends UseCRUDOptionalProps {
  type?: "home" | "delivery";
}

export function useTeam(props?: UseTeamProps) {
  const {
    items,
    loading,
    error,
    reload,
    createItem,
    retrieveItem,
    updateItem,
    deleteItem,
  } = useCRUD<Team, MutateTeam>({
    path: "/api/org/teams",
    ...props,
  });

  const filteredTeams = props?.type
    ? items.filter((team) => team.type === props?.type)
    : items;

  return {
    items: filteredTeams,
    loading,
    error,
    reload,
    createItem,
    retrieveItem,
    updateItem,
    deleteItem,
  };
}
